File: packages/application/src/create-client.use-case.ts
import { Client, type ClientProps, type IClientRepository, type ILogger } from '@monolegal/domain';

export interface CreateClientInput {
  id: string;
  name: string;
  email: string;
}

export class CreateClientUseCase {
  constructor(
    private readonly clientRepository: IClientRepository,
    private readonly logger: ILogger,
  ) {}

  async execute(input: CreateClientInput): Promise<Client> {
    const props: ClientProps = {
      id: input.id,
      name: input.name,
      email: input.email,
    };
    Client.create(props);

    const client = await this.clientRepository.create(props);
    this.logger.info('Client created', { clientId: client.id });
    return client;
  }
}

File: backend/dotnet/src/Monolegal.Application/CreateClientUseCase.cs
using Monolegal.Domain.Entities;
using Monolegal.Domain.Ports;

namespace Monolegal.Application;

public record CreateClientInput(string Id, string Name, string Email);

public class CreateClientUseCase(IClientRepository clientRepository, ILogger logger)
{
    public async Task<Client> ExecuteAsync(CreateClientInput input, CancellationToken cancellationToken = default)
    {
        var props = new ClientProps(input.Id, input.Name, input.Email);
        Client.Create(props);
        var client = await clientRepository.CreateAsync(props, cancellationToken);
        logger.Info("Client created", new Dictionary<string, object?> { ["clientId"] = client.Id });
        return client;
    }
}
